### PR TITLE
Add timestep distribution visualization and loss weighting functionality

### LINKED ISSUE #2383: [Request]  add display "discrete_flow_shift", "timestep_sampling"
優先度が低い要望なのですが、
logger INFOとして、timestep_sampling, discrete_flow_shiftの設定値情報を追加していただけないでしょうか。

# purpose / 目的
discrete_flow_shiftを設定しても、timestep_samplingが対応していない限り、
discrete_flow_shiftが作用しないという状況を可視化し、デバッグしやすくするためです。

# background / 背景
（読み飛ばしていただいて構いません）

* 某LoRA学習系wiki複数において、
Anima登場に合わせて「難しい説明は省きますが、discrete_flow_shift=1.0を調整したほうがいい」という主旨の説明が追加されました。
そこそこ重要な設定値としての位置づけです。
* また、その他サイトなどでも「discrete_flow_shift」の設定値について議論されている際、
「timestep_sampling」の設定が書かれていないことがあり、
そもそも「discrete_flow_shift」の設定が有効になっていない可能性があり、議論が成立していない可能性を憂慮しています。

* 私はそのwiki追記の流れに加担したという立場もあり、
そのwikiを見た方々が、ちゃんとその値が機能しているのかどうか混乱させないようにさせたいです。
wikiの流れとして、従来のflow matchingモデル未経験で、Anima学習に初めて挑戦、またはSDXLからの移行組が多いので、
設定が機能しているかどうか、確かめる能力に慣れていない人が比較的多いと思います。

* animaのshift関連の計算はflux_train_utils.pyを参照しているという点でも（合っていますでしょうか？）、
そこそこpythonの知識がないと発見しにくく、
実はdiscrete_flow_shiftが有効でなかった沼にハマる潜在的な心配はあると思います。
* デフォルト設定値流用するユーザーにとっても、sd-scriptsバージョン更新に合わせたdefault値変化により、
「なぜか、生成結果が明らかに変化した」という混乱を防ぐ効果もあると思います

以上は本来、sd-scriptsが担うべきというわけでもなく（察するに自己責任や派生GUI側の領域でもあり）、
今回の提案は必要というほどでもありませんが、
もし、機会がありましたら追加していただきたく、よろしくお願い致します。

### PR DESCRIPTION
**Pull request by kohya-ss**
This pull request introduces new features and documentation to make timestep sampling and loss weighting more transparent and user-friendly in both FLUX and Anima training workflows. It adds command-line options to visualize the actual timestep distributions and loss weighting, improves logging of sampling configurations, and updates the documentation with clear explanations and usage instructions. The changes affect both the training scripts and the shared utility libraries.

The most important changes are:

**New features: Timestep distribution visualization**
- Added `--show_timesteps` and `--show_timesteps_resolution` arguments to all relevant training scripts via `library/args.py`. These options allow users to visualize the sampled timestep distribution and loss weighting for the current settings, either as an ASCII histogram or a matplotlib image, and then exit without running training.
- Implemented `show_timesteps` functions in both `library/flux_train_utils.py` and `library/anima_train_utils.py`, which generate the visualizations using dummy latents and the current configuration, supporting both FLUX and Anima. (F1359666L531R531, F8df2259L159R159)

**Improved logging and config clarity**
- Added `log_timestep_sampling_info` and `get_timestep_sampling_info` functions to `library/flux_train_utils.py`, and integrated them into all training entrypoints to log a concise, human-readable summary of the timestep sampling configuration at the start of every run, making it clear when options like `--discrete_flow_shift` are ignored. (F1359666L531R531, [[1]](diffhunk://#diff-4a45b918d69f7d16399df9cd8f1a7a5ab4be83636518aea11e9722c94873896aR56-R57) [[2]](diffhunk://#diff-232e93c0cc72afe33c04180ca7e7f04872f6b8046744a443a3eb144321d1f9e6R196-R197) [[3]](diffhunk://#diff-f6eb0889f62391821e5daa01901c95d12b4e4ecf7849b068abdedc28565fd321R47-R48) [[4]](diffhunk://#diff-4e0abf5ef9091e11f3b4691605eea2beda82e3a52e066114d1a8e6a0ec11c5bbR69-R70) [[5]](diffhunk://#diff-1f051f9a3027926ef70b90861593cd22c3eab2be5e2e67f6e90c55fadc2d60f6R75-R76) [[6]](diffhunk://#diff-61545348eb7c86c692d6d966f2889eb419b1e4326b41e48fde2c5818d127382cR52-R53)

**Script and control flow updates**
- Updated all main training scripts (`anima_train.py`, `anima_train_control_net_lllite.py`, `anima_train_network.py`, `flux_train.py`, `flux_train_control_net.py`, `flux_train_network.py`) to check for `--show_timesteps` at startup: if set, the script visualizes the distribution and exits; otherwise, it proceeds with training as usual. [[1]](diffhunk://#diff-4a45b918d69f7d16399df9cd8f1a7a5ab4be83636518aea11e9722c94873896aR768-R770) [[2]](diffhunk://#diff-232e93c0cc72afe33c04180ca7e7f04872f6b8046744a443a3eb144321d1f9e6R822-R824) [[3]](diffhunk://#diff-f6eb0889f62391821e5daa01901c95d12b4e4ecf7849b068abdedc28565fd321R455-R457) [[4]](diffhunk://#diff-4e0abf5ef9091e11f3b4691605eea2beda82e3a52e066114d1a8e6a0ec11c5bbR860-R862) [[5]](diffhunk://#diff-1f051f9a3027926ef70b90861593cd22c3eab2be5e2e67f6e90c55fadc2d60f6R894-R896) [[6]](diffhunk://#diff-61545348eb7c86c692d6d966f2889eb419b1e4326b41e48fde2c5818d127382cR551-R553)

**Documentation improvements**
- Updated both English and Japanese documentation for Anima and FLUX to explain the new visualization options, clarify when `--discrete_flow_shift` is applied or ignored, and provide usage examples. [[1]](diffhunk://#diff-864e35feed5462027a39ccd1cd7b75d7dd77db058ff8b5562b9059e68a0278b7R503-R515) [[2]](diffhunk://#diff-864e35feed5462027a39ccd1cd7b75d7dd77db058ff8b5562b9059e68a0278b7R547-R554) [[3]](diffhunk://#diff-b3b31fc78dd4fba5c6c596dd1ed3bec2ed56280a510d45b7f7645421e52488c1R340-R343) [[4]](diffhunk://#diff-b3b31fc78dd4fba5c6c596dd1ed3bec2ed56280a510d45b7f7645421e52488c1R376-R377)

**Internal refactoring and utility enhancements**
- Refactored utility functions for device selection and latent size parsing to support the new features, and ensured consistent handling of resolution-dependent sampling across both FLUX and Anima. [[1]](diffhunk://#diff-3c2501723a6aff6f17f47c2cfa240729d7f616bc903fead7fea220a40f830e47R4) [[2]](diffhunk://#diff-036aaf8b8965e66f50c6df08b620d3c9fcc7c92e230574c4a9ca0d3b8bf1f1c4R10) [[3]](diffhunk://#diff-3c2501723a6aff6f17f47c2cfa240729d7f616bc903fead7fea220a40f830e47R534-R620)

These changes make it much easier for users to understand and debug how their timestep sampling and loss weighting settings are actually applied during training.

---

Closes #2383.

### CODE REVIEW

File: library/timestep_visualization.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    for _ in tqdm(range(max(1, n_try // batch_size)), desc="sampling timesteps"):
+        timesteps = sample_timesteps(batch_size).detach().flatten().float().cpu().numpy()
+        indices = np.clip(timesteps.astype(np.int64), 0, num_train_timesteps - 1)
+        np.add.at(sampled_timesteps, indices, 1.0)
+
+    # --- compute the loss weighting per timestep ---
+    sampled_weighting = np.ones(num_train_timesteps, dtype=np.float64)
+    if compute_weighting is not None:
+        all_timesteps = torch.arange(1, num_train_timesteps + 1, dtype=torch.float32)
+        weighting = compute_weighting(all_timesteps)
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply library/timestep_visualization.py:66]: _show_console does not depend on the value of timesteps, but only on the position. Therefore, the 1000 items from 1 to 1000 are correctly distributed 25 to each row.

File: library/flux_train_utils.py [hunk 2]
```diff
@@ -530,6 +531,93 @@ def get_noisy_model_input_and_timesteps(
...
+    noise_scheduler = sd3_train_utils.FlowMatchEulerDiscreteScheduler(
+        num_train_timesteps=num_train_timesteps, shift=args.discrete_flow_shift
+    )
+    # latent size for the assumed image resolution (flux_shift reads h, w for the packed-size shift)
+    h, w = parse_show_timesteps_latent_size(args)
+    device, dtype = device_utils.get_preferred_device(), torch.float32
+
+    def sample_timesteps(bsz):
+        latents = torch.zeros(bsz, 16, h, w, dtype=dtype, device=device)
+        noise = torch.ones_like(latents)
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply library/flux_train_utils.py:603]: I'd like to keep a real number of channels for latents.

File: docs/flux_train_network.md [hunk 3]
```diff
@@ -337,6 +337,10 @@ The `--timestep_sampling` option specifies how timesteps (0-1) are sampled:
 
 `--discrete_flow_shift` only applies when `--timestep_sampling` is set to `shift`.
 
+#### Visualizing the Timestep Distribution
+
+You can visualize the actual sampled-timestep distribution and loss weighting for your current settings — and confirm whether `--discrete_flow_shift` is actually applied — using the `--show_timesteps` and `--show_timesteps_resolution` options. A one-line summary of the configuration is also logged at the start of every training run. These options are shared between FLUX and Anima training; see the [Anima guide](anima_train_network.md#visualizing-the-timestep-distribution) for details.
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply docs/flux_train_network.md:342]: ~~I will fix this.~~ This comment appears to be incorrect, as it misidentifies the parts that were changed.

File: library/anima_train_utils.py [hunk 4]
```diff
@@ -158,6 +159,40 @@ def compute_loss_weighting_for_anima(weighting_scheme: str, sigmas: torch.Tensor
...
+    num_train_timesteps = 1000
+    noise_scheduler = sd3_train_utils.FlowMatchEulerDiscreteScheduler(
+        num_train_timesteps=num_train_timesteps, shift=args.discrete_flow_shift
+    )
+    h, w = flux_train_utils.parse_show_timesteps_latent_size(args)  # latent size for the assumed image resolution
+    device, dtype = device_utils.get_preferred_device(), torch.float32
+
+    def sample_timesteps(bsz):
+        latents = torch.zeros(bsz, 16, h, w, dtype=dtype, device=device)
+        noise = torch.ones_like(latents)
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply library/anima_train_utils.py:178]: Same as above.

File: library/flux_train_utils.py [hunk 5]
```diff
@@ -530,6 +531,93 @@ def get_noisy_model_input_and_timesteps(
...
+# ("uniform", "sigmoid", "flux_shift") ignore discrete_flow_shift entirely. This is shared by both
+# FLUX and Anima since Anima reuses get_noisy_model_input_and_timesteps below.
+_SHIFT_AWARE_TIMESTEP_SAMPLING = ("sigma", "shift")
+
+
+def get_timestep_sampling_info(args) -> str:
+    """One-line, human-readable summary of the timestep sampling config.
+
+    Makes it explicit whether ``--discrete_flow_shift`` actually affects the distribution for the chosen
+    ``--timestep_sampling`` (the core confusion behind issue #2383).
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply library/flux_train_utils.py:545]: I think that test would actually be difficult to implement.
